bmcd/src/legacy.rs

```diff
@@ -0,0 +1,395 @@
+//! Routes for legacy API present in versions <= 1.1.0 of the firmware.
+
+use std::fmt::Display;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use actix_web::{rt, web, HttpResponse};
+use nix::sys::statfs::statfs;
+use serde_json::json;
+use tokio::sync::{mpsc, Mutex};
+use tpi_rs::app::bmc_application::{BmcApplication, UsbConfig};
+use tpi_rs::middleware::{NodeId, UsbMode, UsbRoute};
+
+type Query = web::Query<std::collections::HashMap<String, String>>;
+
+trait ToHttpResponse {
+    fn to_response(&self, action: &'static str) -> HttpResponse;
+}
+
+impl<T, E: Display> ToHttpResponse for Result<T, E> {
+    fn to_response(&self, action: &'static str) -> HttpResponse {
+        match self {
+            Ok(_) => success_response(),
+            Err(e) => {
+                let msg = format!("Failed to {}: {}", action, e);
+                HttpResponse::InternalServerError().body(msg)
+            }
+        }
+    }
+}
+
+fn success_response() -> HttpResponse {
+    // For backwards compatibility, `response` holds an array, even though it's unnecessary
+    let msg = json!({
+        "response": [{ "result": "ok" }]
+    });
+
+    HttpResponse::Ok().json(msg)
+}
+
+pub fn config(cfg: &mut web::ServiceConfig) {
+    cfg.service(
+        web::scope("/api/bmc").service(
+            web::resource("")
+                .route(web::get().to(api_entry))
+                .route(web::post().to(api_post)),
+        ),
+    );
+}
+
+// TODO: BmcApplication::new() needs to return just BmcApplication, not wrapped in Arc<>.
```

Comment:
It actually does need to be wrapped in an `Arc` as the `BmcApplication` is shared over multiple threads. e.g. the EventListener is spawned in a new task which holds a ref onto `BmcApplication`
```suggestion

```

Reply:
I overlooked that `BmcApplication` does everything through internal mutability. But it's still not ideal that constructor returns `Arc<Self>`, not `Self`: an Arc should be created at the initialization site, thankfully `BmcApplication` does not need to store `EventListener`. Because otherwise, we create a double Arc (the `actix_web::web::Data` uses `Arc` internally).

What I see is that we should (in the future):

```rust
fn main() {
    let bmc = Arc::new(BmcApplication::new().await?); // new() now returns Self
    bmc.run_event_listener();
    let data = web::Data::from(bmc);

    HttpServer::new(
        .app_data(bmc.clone())
        // ..
}
```

Reply:
This `EventListener` should be moved to a better place, eventually to demonstrate that it's a user of the `BmcApplication` not a part of it. The `Arc` is a leftover of technical debt that is already resolved. I suspect the `EventListener` required a mut `BmcApplication` at that time.

we will update it along the way